src/instructlab/sdg/checkpointing.py

```diff
@@ -0,0 +1,104 @@
+# Standard
+import logging
+import uuid
+
+# Third Party
+from datasets import Dataset, concatenate_datasets, load_dataset
+from datasets.data_files import EmptyDatasetError
+
+logger = logging.getLogger(__name__)
+
+
+class Checkpointer:
+    def __init__(self, checkpoint_dir=None, save_freq=1):
+        self._checkpoint_dir = checkpoint_dir
+
+        self._save_freq = save_freq
+        self._cache = []
+
+    def checkpoint(self, dataset):
+        self._cache.append(dataset)
+        if len(self._cache) < self._save_freq:
+            return
+        self.save()
+        self._cache.clear()
+
+    def done(self):
+        if self._cache:
+            self.save()
+            self._cache.clear()
+
+    def save(self):
+        # TODO: make this dir
+        checkpoint_id = uuid.uuid4().hex
+        checkpoint_file = (
+            f"{self._checkpoint_dir}/data_checkpoint_{checkpoint_id}.jsonl"
+        )
+        logger.info(f"Saving checkpoint to {checkpoint_file}")
+        # Saves all the current records to new file in the checkpoint dir
+        concatenate_datasets(self._cache).to_json(
+            checkpoint_file, orient="records", lines=True
+        )
+
+    def load(self, dataset: Dataset) -> Dataset:
+        # check if _checkpoint_dir exists
+        seed_data = dataset
+        pre_generated_data = None
+        if self._checkpoint_dir is not None:
+            try:
+                pre_generated_data = load_dataset(
+                    "json", data_dir=self._checkpoint_dir, split="train"
+                )
+                logger.info(
+                    f"Loading existing checkpoints from {self._checkpoint_dir}, with {pre_generated_data.num_rows} rows"
+                )
+                seed_data = self._get_missing_data(dataset, pre_generated_data)
+                logger.info(f"Found {seed_data.num_rows} missing rows in the dataset")
+
+            except EmptyDatasetError:
```

Comment:
What raises this? Is it only `load_dataset()`? If so, let's move everything out of this try/except block

Reply:
I beleive so yes, will do